File: firmware/leds.py
from neopixel import Neopixel
import uasyncio as asyncio

class Leds:
    def __init__(self, numpix, pin, brightness=100):
        self.numpix = numpix
        self.neopixel = Neopixel(numpix, 0, pin, "RGB")
        self.neopixel.brightness(brightness)
        self.firstPixelHue = 0

    def update(self):
        for i in range(self.numpix):
            pixelHue = self.firstPixelHue + (i * 65536 // self.numpix)
            self.neopixel.set_pixel(i, self.neopixel.colorHSV(pixelHue, 255, 255))
        self.neopixel.show()
        self.firstPixelHue += 256
        if self.firstPixelHue >= 5 * 65536:
            self.firstPixelHue = 0

    

File: firmware/leds.cpp
#include <Adafruit_NeoPixel.h>
#include "leds.h"

Leds::Leds(int pin, int numPixels) : pixels(numPixels, pin, NEO_RGB + NEO_KHZ800) {
}

void Leds::setup() {
    pixels.begin();
}

void Leds::update() {
    if (firstPixelHue < 5 * 65536) {
        for (int i = 0; i < 13; i++) {
            int pixelHue = firstPixelHue + (i * 65536L / pixels.numPixels());
            pixels.setPixelColor(i, pixels.gamma32(pixels.ColorHSV(pixelHue)));
        }

        pixels.show();
        firstPixelHue += 256;
    } else {
        firstPixelHue = 0;
    }
}